### PR TITLE
Include `Logging` as part of `Create` and `Update` flow

### PR DESCRIPTION
**Pull request by RedbackThomson**
Adds custom hooks to detect changes to the `Logging` spec field and optionally calls the `PutBucketLogging` SDK method. Currently errors are not populated into terminal conditions, and updates are not supported.

Uses https://github.com/aws-controllers-k8s/code-generator/pull/139 and https://github.com/aws-controllers-k8s/code-generator/pull/140

### CODE REVIEW

File: pkg/resource/bucket/hook.go
Comment (vijtrip2) [inline pkg/resource/bucket/hook.go:162]: Why is this `DeepCopy` needed? I noticed no changes being made in `ko` object.

File: pkg/resource/bucket/hook.go
Comment (vijtrip2) [inline pkg/resource/bucket/hook.go:52]: `latest` and `desired` here will be of different types, correct?

It would not cause trouble to compare their primitive fields but shouldn't they both be `ko.Spec.Logging`? Currently `latest` is type from aws-sdk-go and not ACK.

File: Makefile
Comment (jaypipes) [inline Makefile:26]: Consider putting these changes into a totally separate PR, as it's not germane to this PR.

**Review by jaypipes — COMMENTED** [review]
@RedbackThomson please rebase this to remove the unrelated 0.7.1 runtime update changes...

File: apis/v1alpha1/types.go
Comment (jaypipes) [inline apis/v1alpha1/types.go:49]: ugh. a field called "Enabled" that isn't a boolean and is a struct? :(

File: pkg/resource/bucket/hook.go
Comment (jaypipes) [reply pkg/resource/bucket/hook.go:52]: Agreed. @RedbackThomson we typically use the variable name `resp` for the response object from the SDK call. Recommend using `resp` here.

File: pkg/resource/bucket/hook.go
Comment (jaypipes) [inline pkg/resource/bucket/hook.go:57]: instead of creating a separate delta object here, you could change this function signature to accept a delta object and place a call to `diffLogging` in the `delta_pre_compare` hook:

https://github.com/aws-controllers-k8s/code-generator/blob/cae7cb2cccda567eab55c01873f89274c8f82b60/templates/pkg/resource/delta.go.tpl#L22-L24

and pass in the `delta` variable from there.

File: pkg/resource/bucket/hook.go
Comment (jaypipes) [inline pkg/resource/bucket/hook.go:31]: instead of doing this, see my comment below about adding the `diffLogging` call in the `delta_pre_compare` hook and passing the `delta` object to this function. That way, here you can simply do:

```go
if delta.DifferentAt("Spec.Logging")
```

File: pkg/resource/bucket/hook.go
Comment (jaypipes) [reply pkg/resource/bucket/hook.go:162]: @vijtrip2 Nick is simply copy/pasting from the normal sdk_update template... nothing really wrong with that... though techincally he should be doing if **below** the call to `rm.sdkapi.PutBucketLogging`

**Review by jaypipes — CHANGES_REQUESTED** [review]
Nick, I think you can use the `delta_pre_compare` hook point to simplify your code a bit and remove the need for (re-)creating delta objects.

**Comment by RedbackThomson** [comment]
Thank you @vijtrip2 and @jaypipes for your reviews. I have updated the code-generator to remove a lot of the heavy-lifting for this - namely creating the `Put*` payload objects from `spec`. I have removed the `diff` method, as well, because I would instead prefer to update the fields of a `resource` type and then use the (already generated) `diffResource` code. However, that will be another pull request for supporting update operations.

**Comment by RedbackThomson** [comment]
Oops, I added update as well because it wasn't that much extra code. Sorry :)

File: templates/hooks/bucket_sdk_file_end.go.tpl
Comment (jaypipes) [inline templates/hooks/bucket_sdk_file_end.go.tpl:25]: nit: missing space between the `}}` and the word `end`

File: templates/hooks/bucket_sdk_file_end.go.tpl
Comment (jaypipes) [inline templates/hooks/bucket_sdk_file_end.go.tpl:21]: nit: add a space between `}}` and `1`

File: templates/hooks/bucket_sdk_file_end.go.tpl
Comment (jaypipes) [inline templates/hooks/bucket_sdk_file_end.go.tpl:14]: recommend calling this `new{{ $memberRefName }}` instead, since it's not creating (in terms of a Create API call) and you call these methods in the `newGetBucketLoggingPayload` method...

**Review by jaypipes — CHANGES_REQUESTED** [review]
Looking very good... couple teeny nits inline.

**Comment by RedbackThomson** [comment]
/hold 

I think this is ready to go, but want to wait for https://github.com/aws-controllers-k8s/code-generator/pull/140 before merging.

**Review by jaypipes — APPROVED** [review]
:+1:

**Pull request merged**
